codalab/worker/dependency_manager.py

```diff
@@ -75,110 +124,125 @@ def __init__(
+                    f'State file did not exist. Creating one at path {self._state_committer.path}.'
                 )
-            )
-            del self._dependencies[dep]
-            del self._dependency_locks[dep]
 
-        # Remove the orphaned directories from the local file system
-        directories_to_remove = local_directories - self._paths
-        for dir in directories_to_remove:
-            full_path = os.path.join(self.dependencies_dir, dir)
-            logger.info(
-                "Remove orphaned directory {} from the local file system.".format(full_path)
-            )
-            remove_path(full_path)
+            # Get the paths that exist in dependency state, loaded path and
+            # the local file system (the dependency directories under self.dependencies_dir)
+            local_directories = set(os.listdir(self.dependencies_dir))
+            paths_in_loaded_state = [dep_state.path for dep_state in dependencies.values()]
+            paths = paths.intersection(paths_in_loaded_state).intersection(local_directories)
+
+            # Remove the orphaned dependencies if they don't exist in paths
+            # (intersection of paths in dependency state, loaded paths and the paths on the local file system)
+            dependencies_to_remove = [
+                dep for dep, dep_state in dependencies.items() if dep_state.path not in paths
+            ]
+            for dep in dependencies_to_remove:
+                logger.info(
+                    "Dependency {} in dependency state but its path {} doesn't exist on the local file system. "
+                    "Removing it from dependency state.".format(
+                        dep, os.path.join(self.dependencies_dir, dependencies[dep].path)
+                    )
+                )
+                del dependencies[dep]
+
+            # Remove the orphaned directories from the local file system
+            directories_to_remove = local_directories - paths
+            for directory in directories_to_remove:
+                full_path = os.path.join(self.dependencies_dir, directory)
+                if os.path.exists(full_path):
+                    logger.info(
+                        "Remove orphaned directory {} from the local file system.".format(full_path)
+                    )
+                    remove_path(full_path)
+
+            # Save the current synced state back to the state file: dependency-state.json as
+            # the current state might have been changed during the state syncing phase
+            self._commit_state(dependencies, paths)
+
+    def _fetch_state(self, default=None):
+        """
+        Fetch state from JSON file stored on disk.
+        NOT NFS-SAFE - Caller should acquire self._state_lock before calling this method.
+        WARNING: If a value for `default` is specified, errors will be silently handled.
+        """
+        state: DependencyManagerState = self._state_committer.load(default)
```

Comment:
Add an assertion at the beginning of this method (and all non-NFS-safe methods) that self._state_lock is held